[BE] fix: 질문을 position 순서대로 내려주도록 로직 수정

nayonsoso opened the pull request:
<!-- 제목: [BE/FE/All] (기능 등) -->
<!-- 아래에 이슈 번호를 매겨주세요 -->

- resolves #831 

---

### 🚀 어떤 기능을 구현했나요 ?
- 질문이 position 순으로 정렬되지 않는 아래의 이슈를 해결했습니다. 
![image](https://github.com/user-attachments/assets/4329da6a-96f7-449a-8470-bbc827f9978f)


### 🔥 어떻게 해결했나요 ?
- AS-IS: Question을 position으로 정렬한 값을 받아서 Map<Question, List<Answer>> 으로 만든 다음, mapper 에 전달했다. 이때, map 은 순서를 보장하지 않으므로 mapper 에서 position 순서대로 내려지지 않았다.
- TO-BE: 쿼리에서 정렬하지 않고, mapper에서 정렬하도록 바꿔주었다.

### 📝 어떤 부분에 집중해서 리뷰해야 할까요?
- 

### 📚 참고 자료, 할 말

nayonsoso committed:
fix: 질문을 순서대로 내려주는 로직 수정

AS-IS: Question을 position으로 정렬한 값을 받아서 Map<Question, List<Answer>> 으로 만든 다음, mapper 에 전달했다. 이때, map 은 순서를 보장하지 않으므로 mapper 에서 position 순서대로 내려지지 않았다.
TO-BE: 쿼리에서 정렬하지 않고, mapper에서 정렬하도록 바꿔주었다.

donghoony commented on backend/src/main/java/reviewme/review/service/ReviewGatheredLookupService.java:
```diff
@@ -54,7 +54,7 @@ private Section getSectionOrThrow(long sectionId, ReviewGroup reviewGroup) {
-                .findAllBySectionIdOrderByPosition(section.getId())
+                .findAllBySectionId(section.getId())
                 .stream()
                 .collect(Collectors.toMap(Question::getId, Function.identity()));
```
`LinkedHashMap`으로 받아왔으면 추가 구현 없이 정렬되지 않았을까요 ?

nayonsoso replied:
음~ 맛있다😋

donghoony reviewed (changes requested):
한 번만 확인하고, 더 쉬운 방법이 있다면 이쪽으로 가는 것도 좋겠습니다 👍🏻

nayonsoso committed:
Revert "fix: 질문을 순서대로 내려주는 로직 수정"

This reverts commit 4ff8cc37e079b5aaa2d038ffa884d94fbefecbdd.

nayonsoso committed:
fix: 질문을 순서대로 내려주는 로직 수정

AS-IS: Question을 position으로 정렬한 값을 받아서 Map<Question, List<Answer>> 으로 만든 다음, mapper 에 전달했다. 이때, map 은 순서를 보장하지 않으므로 mapper 에서 position 순서대로 내려지지 않았다.
TO-BE: LinkedHashMap 을 사용해서 순서를 보장한다.

donghoony commented on backend/src/main/java/reviewme/review/service/ReviewGatheredLookupService.java:
```diff
@@ -53,10 +54,11 @@ private Section getSectionOrThrow(long sectionId, ReviewGroup reviewGroup) {
-                .stream()
-                .collect(Collectors.toMap(Question::getId, Function.identity()));
+        List<Question> questions = questionRepository.findAllBySectionIdOrderByPosition(section.getId());
+        LinkedHashMap<Long, Question> questionIdQuestion = new LinkedHashMap<>();
+        for (Question question : questions) {
+            questionIdQuestion.put(question.getId(), question);
+        }
```
```suggestion
        Map<Long, Question> questionIdQuestion = new LinkedHashMap<>();
        questions.forEach(question -> questionIdQuestion.put(question.getId(), question));
```

nayonsoso replied:
음 맛있다🤤 반영완했습니다!

donghoony reviewed (approved):
어떤가요?

nayonsoso committed:
refactor: 가독성 개선

Kimprodp commented on backend/src/main/java/reviewme/review/service/ReviewGatheredLookupService.java:
```diff
@@ -53,10 +54,9 @@ private Section getSectionOrThrow(long sectionId, ReviewGroup reviewGroup) {
-                .collect(Collectors.toMap(Question::getId, Function.identity()));
+        List<Question> questions = questionRepository.findAllBySectionIdOrderByPosition(section.getId());
+        Map<Long, Question> questionIdQuestion = new LinkedHashMap<>();
+        questions.forEach(question -> questionIdQuestion.put(question.getId(), question));
```
동일한 맥락으로 questionIdAnswers 도 순서가 보장되지 않을 수 있지 않나요?

nayonsoso replied:
WoW... 마지막에 리턴을 할 때 그냥 map 으로 collect 하기 때문에 순서가 보장되지 않을 수도 있었네요!!
꼼꼼테드 고맙습니다🥹 테스트 코드도 보강했어요!

Kimprodp reviewed (changes requested):
답변만 확인해주십셔!

hyeonji1220 commented on backend/src/main/java/reviewme/review/service/ReviewGatheredLookupService.java:
```diff
@@ -53,10 +54,9 @@ private Section getSectionOrThrow(long sectionId, ReviewGroup reviewGroup) {
+        Map<Long, Question> questionIdQuestion = new LinkedHashMap<>();
+        questions.forEach(question -> questionIdQuestion.put(question.getId(), question));
 
         Map<Long, List<Answer>> questionIdAnswers = answerRepository
                 .findReceivedAnswersByQuestionIds(reviewGroup.getId(), questionIdQuestion.keySet(),
```
이 부분도 그렇고 아래의 return 부분 mapping도 그렇고 그냥 linkedHashMap은 아니라 섞일 수 있는 가능성이 있는 거 아닌가 싶은데 현재 테스트 코드에서는 100번씩 repeatableTest 돌려도 정렬돼서 나오네요ㅎㅎ 혹시 원리를 아시는지..

nayonsoso replied:
히히 테드의 리뷰를 받고 반영했습니다~
그리고 "테스트 코드가 통과하는 것"은 사실 테스트 코드도 이상하게 되어있더라고요😅
제가 fixture 로 4개의 Question 을 만들고, 각각의 이름대로 정렬되어있는지를 테스트했었는데,
fixture 는 질문의 내용의 내용이 모두 같았으므로 의미가 없는 테스트였었습니다🥲
(테드 피드백 반영하기 전에 테스트 코드 수정부터 했었는데, 터지는 테스트였더라고요😱)

지금은 테스트도 수정했습니다!

hyeonji1220 reviewed (approved):
산초 `이게 왜 되지?` 질문만 답변 부탁드립니당ㅎㅎ

nayonsoso committed:
refactor: 질문 순서 보장 로직 수정

- 반환값도 LinkedHashMap 이 되도록

hyeonji1220 reviewed (approved):
반영 확인 완! 이제 가지런히 정렬된 페이지를 볼 수 있겠곤요👏

nayonsoso committed:
fix: 컴파일 에러 해결

Kimprodp reviewed (approved):
반완확 고생했어요 신초~
